interpreter/decorators.go

```diff
@@ -76,47 +77,128 @@ func decDisableShortcircuits() InterpretableDecorator {
 	}
 }
 
-// decFoldConstants checks whether the arguments the create list and map operations are all
-// constant values and constructs a new aggregate value. Future improvements to this method will
-// also do the same for typed object creations and functions whose arguments are constant.
-func decFoldConstants() InterpretableDecorator {
+// decOptimize optimizes the program plan by looking for common evaluation patterns and
+// conditionally precomputating the result.
+// - build list and map values with constant elements.
+// - convert 'in' operations to set membership tests if possible.
+func decOptimize() InterpretableDecorator {
 	return func(i Interpretable) (Interpretable, error) {
 		switch i.(type) {
 		case *evalList:
-			l := i.(*evalList)
-			for _, elem := range l.elems {
-				_, isConst := elem.(*evalConst)
-				if !isConst {
-					return i, nil
-				}
-			}
-			val := l.Eval(EmptyActivation())
-			return &evalConst{
-				id:  l.id,
-				val: val,
-			}, nil
+			return maybeBuildListLiteral(i)
 		case *evalMap:
-			mp := i.(*evalMap)
-			for idx, key := range mp.keys {
-				_, isConst := key.(*evalConst)
-				if !isConst {
-					return i, nil
-				}
-				_, isConst = mp.vals[idx].(*evalConst)
-				if !isConst {
-					return i, nil
-				}
+			return maybeBuildMapLiteral(i)
+		case *evalBinary:
+			call := i.(*evalBinary)
+			if call.overload == overloads.InList {
+				return maybeOptimizeSetMembership(i)
 			}
-			val := mp.Eval(EmptyActivation())
-			return &evalConst{
-				id:  mp.id,
-				val: val,
-			}, nil
 		}
 		return i, nil
 	}
 }
 
+func maybeBuildListLiteral(i Interpretable) (Interpretable, error) {
+	l := i.(*evalList)
```

Comment:
Would you mind promoting these back to the `switch` statement so that you receive an `evalList` here? The reason is that once this list grows, it will be easier to make a mistake on casting it incorrectly. This same comment would apply to all `func`s called from the switch statement above.